File: Scripts/Comparison/compare_R_Python_spart_marmit.py
"""Cross-language robustness comparison: SPART and MARMIT (both versions),
ToolsRTM (R) vs toolsrtm (Python).

Reads the LUTs/reflectance written by compare_R_Python_spart_marmit.R,
runs the matching Python simulation for every row, reports RMSE/R^2/bias
per model, and produces comparison figures.
"""
import os
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

from toolsrtm import spart_toa, sentinel2a_msi, get_marmit_rsoil

HERE = os.path.dirname(os.path.abspath(__file__))
OUTDIR = os.path.join(HERE, "_out")


def stats(R, Py):
    diff = Py - R
    rmse = float(np.sqrt(np.mean(diff ** 2)))
    max_abs = float(np.max(np.abs(diff)))
    bias = float(np.mean(diff))
    ss_res = np.sum(diff ** 2)
    ss_tot = np.sum((R - np.mean(R)) ** 2)
    r2 = float(1 - ss_res / ss_tot) if ss_tot > 0 else np.nan
    return dict(rmse=rmse, max_abs=max_abs, bias=bias, r2=r2)


# --- 1. SPART ---
lut = pd.read_csv(os.path.join(OUTDIR, "lut_SPART.csv"))
toc_R = pd.read_csv(os.path.join(OUTDIR, "refl_R_SPART_toc.csv"))
toa_R = pd.read_csv(os.path.join(OUTDIR, "refl_R_SPART_toa.csv"))
bands_df = pd.read_csv(os.path.join(OUTDIR, "SPART_bands.csv"))
band_cols = [c for c in toc_R.columns if c.startswith("band")]
n = len(lut)

sensor = sentinel2a_msi()
toc_py = np.full((n, len(band_cols)), np.nan)
toa_py = np.full((n, len(band_cols)), np.nan)
for i in range(n):
    row = lut.iloc[i]
    inputLUT = {k: row[k] for k in lut.columns}
    try:
        res = spart_toa(inputLUT, sensor=sensor, leaf_model="PROSPECT-PRO",
                         BSMBrightness=row["BSMBrightness"], BSMlat=row["BSMlat"],
                         BSMlon=row["BSMlon"], SMp=row["SMp"])
        if len(res.rfl_toc_brdf) == len(band_cols) and np.all(np.isfinite(res.rfl_toc_brdf)):
            toc_py[i, :] = res.rfl_toc_brdf
            toa_py[i, :] = res.rfl_toa
    except Exception:
        pass

R_toc = toc_R[band_cols].to_numpy()
R_toa = toa_R[band_cols].to_numpy()
ok = (~np.isnan(toc_py).any(axis=1)) & (~np.isnan(R_toc).any(axis=1)) & \
     (~np.isnan(toa_py).any(axis=1)) & (~np.isnan(R_toa).any(axis=1))
n_ok = int(ok.sum())
toc_stats = stats(R_toc[ok], toc_py[ok])
toa_stats = stats(R_toa[ok], toa_py[ok])
print(f"SPART TOC-BRDF: {n_ok}/{n} valid, R2={toc_stats['r2']:.8f} RMSE={toc_stats['rmse']:.3e}")
print(f"SPART TOA     : {n_ok}/{n} valid, R2={toa_stats['r2']:.8f} RMSE={toa_stats['rmse']:.3e}")

spart_summary = pd.DataFrame([
    dict(product="TOC-BRDF", n_valid=n_ok, **toc_stats),
    dict(product="TOA", n_valid=n_ok, **toa_stats),
])
spart_summary.to_csv(os.path.join(OUTDIR, "comparison_spart_summary.csv"), index=False)

wave = bands_df["wave_nm"].to_numpy()
fig, axes = plt.subplots(1, 3, figsize=(15, 4.2))
axes[0].scatter(R_toc[ok].mean(axis=1), toc_py[ok].mean(axis=1), s=10, alpha=0.6, color="#2166AC")
lims = [min(R_toc[ok].min(), toc_py[ok].min()), max(R_toc[ok].max(), toc_py[ok].max())]
axes[0].plot(lims, lims, "--", color="grey")
axes[0].set_xlabel("R (ToolsRTM) TOC BRDF"); axes[0].set_ylabel("Python (toolsrtm) TOC BRDF")
axes[0].set_title(f"TOC, mean/sim\nR2={toc_stats['r2']:.6f}")

axes[1].scatter(R_toa[ok].mean(axis=1), toa_py[ok].mean(axis=1), s=10, alpha=0.6, color="#B2182B")
lims2 = [min(R_toa[ok].min(), toa_py[ok].min()), max(R_toa[ok].max(), toa_py[ok].max())]
axes[1].plot(lims2, lims2, "--", color="grey")
axes[1].set_xlabel("R (ToolsRTM) TOA"); axes[1].set_ylabel("Python (toolsrtm) TOA")
axes[1].set_title(f"TOA, mean/sim\nR2={toa_stats['r2']:.6f}")

idx0 = np.where(ok)[0][0]
axes[2].plot(wave, R_toc[idx0], "o-", color="black", label="R TOC (ToolsRTM)")
axes[2].plot(wave, toc_py[idx0], "x--", color="#2166AC", label="Python TOC")
axes[2].plot(wave, R_toa[idx0], "o-", color="grey", label="R TOA (ToolsRTM)")
axes[2].plot(wave, toa_py[idx0], "x--", color="#B2182B", label="Python TOA")
axes[2].set_xlabel("Wavelength (nm)"); axes[2].set_ylabel("Reflectance")
axes[2].set_title(f"Example sim (row {idx0 + 1}), Sentinel-2A bands")
axes[2].legend(fontsize=7)
plt.tight_layout()
plt.savefig(os.path.join(OUTDIR, "comparison_spart.png"), dpi=150)
plt.close(fig)
print("Wrote comparison_spart.png")

# --- 2. MARMIT (both versions) ---
marmit_grid = pd.read_csv(os.path.join(OUTDIR, "lut_MARMIT.csv"))
wl_out = np.arange(400, 2501)
n_i, k_i, d_i = 1.53, 0.001, 0.0005

marmit_results = []
fig, axes = plt.subplots(1, 2, figsize=(11, 4.5))
for ax, version in zip(axes, ["marmit1", "marmit2"]):
    refl_R = pd.read_csv(os.path.join(OUTDIR, f"refl_R_MARMIT_{version}_wet.csv"))
    wl_cols = [c for c in refl_R.columns if c.startswith("wl")]
    R_mat = refl_R[wl_cols].to_numpy()
    nrows = len(marmit_grid)
    py_mat = np.full((nrows, len(wl_cols)), np.nan)
    for i in range(nrows):
        row = marmit_grid.iloc[i]
        try:
            soil = get_marmit_rsoil(soil_id=int(row["soil_id"]), L=row["L"], eps=row["eps"],
                                     version=version, n_i=n_i, k_i=k_i, d_i=d_i, wl_out=wl_out)
            if len(soil.rsoil_wet) == len(wl_cols) and np.all(np.isfinite(soil.rsoil_wet)):
                py_mat[i, :] = soil.rsoil_wet
        except Exception:
            pass
    ok_m = (~np.isnan(py_mat).any(axis=1)) & (~np.isnan(R_mat).any(axis=1))
    n_ok_m = int(ok_m.sum())
    s = stats(R_mat[ok_m], py_mat[ok_m])
    print(f"MARMIT ({version}): {n_ok_m}/{nrows} valid, R2={s['r2']:.8f} RMSE={s['rmse']:.3e}")
    marmit_results.append(dict(version=version, n_valid=n_ok_m, **s))

    idx0 = np.where(ok_m)[0][0]
    ax.plot(wl_out, R_mat[idx0], color="black", linewidth=1.1, label="R (ToolsRTM)")
    ax.plot(wl_out, py_mat[idx0], color="#B2182B", linewidth=1, linestyle="--", label="Python (toolsrtm)")
    ax.set_title(f"{version}, example wet-soil spectrum (row {idx0+1})\nR2={s['r2']:.6f}, n={n_ok_m}/{nrows}")
    ax.set_xlabel("Wavelength (nm)"); ax.set_ylabel("Wet-soil reflectance")
    ax.legend(fontsize=8)

marmit_summary = pd.DataFrame(marmit_results)
marmit_summary.to_csv(os.path.join(OUTDIR, "comparison_marmit_summary.csv"), index=False)
plt.tight_layout()
plt.savefig(os.path.join(OUTDIR, "comparison_marmit.png"), dpi=150)
plt.close(fig)
print("Wrote comparison_marmit.png")
print(marmit_summary.to_string(index=False))

File: Scripts/Comparison/compare_R_Python_spart_marmit.R
## Cross-language robustness comparison: SPART (soil-plant-atmosphere) and
## MARMIT (wet-soil reflectance, both versions) -- ToolsRTM (R) vs
## toolsrtm (Python). Same shared-LUT approach as the other
## Scripts/Comparison/*.R scripts.

this_file <- sub("^--file=", "", grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE))
root <- normalizePath(file.path(dirname(this_file), "..", ".."))
outdir <- file.path(root, "Scripts/Comparison/_out")
dir.create(outdir, showWarnings = FALSE, recursive = TRUE)

suppressMessages(devtools::load_all(file.path(root, "ToolsRTM"), quiet = TRUE))

## ---------------------------------------------------------------------
## 1. SPART: TOC + TOA reflectance, Sentinel-2A bands, realistic fixed
##    atmosphere (matching t03-spart.Rmd's own choice to avoid the known
##    unphysical-atmosphere edge case at getLUT()'s wide default ranges).
## ---------------------------------------------------------------------
n_spart <- 100
LUT_spart <- as.data.frame(getLUT(inputs = ToolsRTM::inputsSPART, nLUT = n_spart, setseed = 21))
LUT_spart$Cs <- 0; LUT_spart$fqe <- 0.01; LUT_spart$Cx <- 0
LUT_spart$Pa <- 1000; LUT_spart$aot550 <- 0.3246; LUT_spart$uo3 <- 0.3480; LUT_spart$uh2o <- 1.4116
LUT_spart$alt_m <- 0; LUT_spart$Pa0 <- 1000
LUT_spart$BSMBrightness <- 0.5; LUT_spart$BSMlat <- 25; LUT_spart$BSMlon <- 45; LUT_spart$SMp <- 15

rfl_toc <- matrix(NA_real_, n_spart, 13)
rfl_toa <- matrix(NA_real_, n_spart, 13)
wave_bands <- NULL
for (i in seq_len(n_spart)) {
  sim <- tryCatch(suppressWarnings(SPART(inputLUT = LUT_spart[i, , drop = FALSE], CanopyModel = "fourSAIL",
                                          LeafModel = "PROSPECT-PRO", sensor.i = ToolsRTM::Sentinel2A.MSI,
                                          rsoil = NULL, get.plots = FALSE)),
                  error = function(e) NULL)
  if (!is.null(sim) && all(is.finite(sim$output$rfl.toc.BRDF)) && all(is.finite(sim$output$rfl.toa))) {
    if (is.null(wave_bands)) wave_bands <- sim$output$wave
    if (length(sim$output$rfl.toc.BRDF) == 13) {
      rfl_toc[i, ] <- sim$output$rfl.toc.BRDF
      rfl_toa[i, ] <- sim$output$rfl.toa
    }
  }
}
ok_spart <- stats::complete.cases(rfl_toc) & stats::complete.cases(rfl_toa)
cat("R side, SPART:", sum(ok_spart), "/", n_spart, "valid\n")

write.csv(LUT_spart, file.path(outdir, "lut_SPART.csv"), row.names = FALSE)
out_toc <- as.data.frame(rfl_toc); colnames(out_toc) <- paste0("band", seq_len(13))
out_toc$row <- seq_len(n_spart); out_toc$ok <- ok_spart
write.csv(out_toc, file.path(outdir, "refl_R_SPART_toc.csv"), row.names = FALSE)
out_toa <- as.data.frame(rfl_toa); colnames(out_toa) <- paste0("band", seq_len(13))
out_toa$row <- seq_len(n_spart); out_toa$ok <- ok_spart
write.csv(out_toa, file.path(outdir, "refl_R_SPART_toa.csv"), row.names = FALSE)
write.csv(data.frame(band = seq_len(13), wave_nm = wave_bands), file.path(outdir, "SPART_bands.csv"), row.names = FALSE)

## ---------------------------------------------------------------------
## 2. MARMIT: both versions (marmit1, marmit2), varying soil id/L/eps,
##    against the bundled Bablet_2016 database (17 soil IDs).
## ---------------------------------------------------------------------
set.seed(22)
n_marmit <- 60
marmit_grid <- data.frame(
  soil_id = sample(1:17, n_marmit, replace = TRUE),
  L = runif(n_marmit, 0.01, 0.15),
  eps = runif(n_marmit, 0.1, 0.9)
)
n_i <- 1.53; k_i <- 0.001; d_i <- 0.0005
wl_out <- 400:2500

run_marmit_version <- function(version) {
  wet <- matrix(NA_real_, n_marmit, length(wl_out))
  dry <- matrix(NA_real_, n_marmit, length(wl_out))
  for (i in seq_len(n_marmit)) {
    soil <- tryCatch(get.marmit.rsoil(database = "Bablet_2016", id = marmit_grid$soil_id[i], version = version,
                                       L = marmit_grid$L[i], eps = marmit_grid$eps[i],
                                       n_i = n_i, k_i = k_i, d_i = d_i, wl.out = wl_out),
                      error = function(e) NULL)
    if (!is.null(soil) && length(soil$rsoil.wet) == length(wl_out)) {
      wet[i, ] <- soil$rsoil.wet
      dry[i, ] <- soil$rsoil.dry
    }
  }
  ok <- stats::complete.cases(wet)
  cat("R side, MARMIT (", version, "):", sum(ok), "/", n_marmit, "valid\n")
  list(wet = wet, dry = dry, ok = ok)
}

write.csv(marmit_grid, file.path(outdir, "lut_MARMIT.csv"), row.names = FALSE)
for (version in c("marmit1", "marmit2")) {
  res <- run_marmit_version(version)
  out_wet <- as.data.frame(res$wet); colnames(out_wet) <- paste0("wl", wl_out)
  out_wet$row <- seq_len(n_marmit); out_wet$ok <- res$ok
  write.csv(out_wet, file.path(outdir, paste0("refl_R_MARMIT_", version, "_wet.csv")), row.names = FALSE)
}

cat("Done. SPART + MARMIT (both versions) written to", outdir, "\n")
